CI: Add PVS-Studio Analysis for Windows

derrod opened the pull request:
### Description

Runs PVS-Studio on Windows builds during scheduled runs.

### Motivation and Context

Static analysis is nice to have, SARIF integrates with GitHub so maintainers can easily manage and action on automatic analysis reports.

Settings this up for Linux and macOS is a bit more involved as we can't just run it on a project file as on Windows, and the command line/configuration files are different. There will be some duplication between the three platforms which may be annoying, so ideally we'd also try to cut non-Windows platforms down to platform-specific files. We'll deal with that at a later date.

### How Has This Been Tested?

Ran on my fork.

### Types of changes

- New feature (non-breaking change which adds functionality)

### Checklist:

- [x] My code has been run through [clang-format](https://github.com/obsproject/obs-studio/blob/master/.clang-format).
- [x] I have read the [**contributing** document](https://github.com/obsproject/obs-studio/blob/master/CONTRIBUTING.rst).
- [x] My code is not on the master branch.
- [x] The code has been tested.
- [x] All commit messages are properly formatted and commits squashed where appropriate.
- [x] I have included updates to all appropriate documentation.

PatTheMav commented on .github/actions/analyse-obs/action.yaml:
```diff
@@ -0,0 +1,60 @@
+    - name: 'Setup PVS Studio'
+      shell: pwsh
+      run: |
+        Invoke-WebRequest -Uri https://cdn.pvs-studio.com/PVS-Studio_setup.exe -OutFile ..\PVS-Studio_setup.exe
```
We should use hash checks here as with all other external code and binaries we download in our CI workflows/actions.

derrod replied:
Unfortunately they don't seem to have versioned download URLs, so this would break every time they change it. So a hash could be set as non-secret actions variable, or we could rely on the code signature (via `Get-AuthenticodeSignature`).

PatTheMav commented on .github/actions/analyse-obs/action.yaml:
```diff
@@ -0,0 +1,60 @@
+            "-o", "${{ github.workspace }}\analysis.plog"
+            "-C", "${env:GITHUB_ACTION_PATH}\obs.pvsconfig"
+        )
+        & "C:\Program Files (x86)\PVS-Studio\PVS-Studio_Cmd.exe" @pvsParams
```
Should use `Invoke-External` here (which includes exit code handling already).

derrod replied:
Since the exit code will be pretty much always non-zero (since 256 means that it found something, see the [documentation](https://pvs-studio.com/en/docs/manual/0035/)) that would always throw, so I'd have to wrap it in a `try...catch` block anyway.

The best thing to do here would probably be to write a little script that actually handles the return codes properly and only throws for those that are actual errors.

PatTheMav commented on .github/actions/analyse-obs/action.yaml:
```diff
@@ -0,0 +1,60 @@
+    - name: Convert Analysis to Sarif
+      shell: pwsh
+      run: |
+        & 'C:\Program Files (x86)\PVS-Studio\PlogConverter.exe' -a GA:1,2 -d V1042,Renew -t Sarif "${{ github.workspace }}\analysis.plog"
```
See above: Use `Invoke-External`.

PatTheMav commented on .github/workflows/build-project.yaml:
```diff
@@ -365,6 +365,15 @@ jobs:
           target: x64
           config: ${{ needs.check-event.outputs.config }}
 
+      - name: Run PVS-Studio Analysis
```
I'd rather have us use a new `analyze-project` workflow that runs in parallel to the build workflow. This one could then be extended to use the clang analyse build or ASAN and other code quality checkers.

PatTheMav reviewed (commented):
I think in general it should be fine to have it run on Windows builds at first and use clang's analysis on Linux and macOS.

derrod commented:
Made the following changes:

- Switched to using chocolatey to install PVS as it allows us to pin a version, additionally it should help with @notr1ch's concerns about CDN reliability as the package will be downloaded from Chocolatey's repository
- Changed to using `Invoke-External` for activation/conversion calls
- Added flags enum with error codes to properly log errors for the main binary, ignoring the code that only signals that code issues were found (`256`)
- Renamed action to `windows-analysis` since this is Windows-specific and Linux/macOS will be separate 
- Split it into `analyze-project` workflow that runs in parallel, also uses `Debug` configuration instead of release

PatTheMav commented on .github/actions/windows-analysis/action.yaml:
```diff
@@ -0,0 +1,88 @@
+      uses: actions/upload-artifact@v4
+      with:
+        name: 'pvs-analysis'
+        path: "${{ github.workspace }}/analysis.*"
```
Are the extensions of the generated files known or are they fully arbitrary (e.g. does PVS add a randomised string to the file names)?

derrod replied:
There's `analysis.plog` and `analysis.plog.sarif` so I simply put in a wildcard to get both.

PatTheMav replied:
Ah suspected as much that it's additional suffixes.

PatTheMav commented on .github/workflows/analyze-project.yaml:
```diff
@@ -0,0 +1,39 @@
+on:
+  workflow_call:
+jobs:
+  windows-analysis:
```
I think "windows" should suffice in this case as the "analysis" part is provided by the workflow name.

If it has to be part of the job name, then I'd prefer "analysis-windows", as it fits better with "analysis-linux" and "analysis-macros" later.

RytoEX commented on .github/actions/windows-analysis/action.yaml:
```diff
@@ -0,0 +1,90 @@
+name: Run PVS-Studio analysis
+inputs:
+  pvsUsername:
+    description: PVS Studio Licence Username
```
```suggestion
    description: PVS-Studio License Username
```

RytoEX commented on .github/actions/windows-analysis/action.yaml:
```diff
@@ -0,0 +1,90 @@
+    description: PVS Studio Licence Username
+    required: true
+  pvsKey:
+    description: PVS Studio Licence Key
```
```suggestion
    description: PVS-Studio License Key
```

RytoEX commented on .github/actions/windows-analysis/action.yaml:
```diff
@@ -0,0 +1,90 @@
+runs:
+  using: composite
+  steps:
+    - name: 'Setup PVS Studio'
```
```suggestion
    - name: 'Setup PVS-Studio'
```

RytoEX commented on .github/actions/windows-analysis/action.yaml:
```diff
@@ -0,0 +1,90 @@
+      run: |
+        choco install pvs-studio --version=7.30.81185.980 -y --no-progress
+
+    - name: Activate PVS Studio
```
```suggestion
    - name: Activate PVS-Studio
```

RytoEX commented on .github/actions/windows-analysis/action.yaml:
```diff
@@ -0,0 +1,90 @@
+        }
+        exit $pvs_result
+
+    - name: Convert Analysis to Sarif
```
```suggestion
    - name: Convert Analysis to SARIF
```

RytoEX commented on .github/actions/windows-analysis/action.yaml:
```diff
@@ -0,0 +1,90 @@
+          ${{ github.workspace }}/analysis.plog
+          ${{ github.workspace }}/analysis.plog.sarif
+
+    - name: 'Upload PVS report'
```
```suggestion
    - name: 'Upload PVS Report'
```

RytoEX commented on .github/workflows/analyze-project.yaml:
```diff
@@ -0,0 +1,39 @@
+  workflow_call:
+jobs:
+  windows:
+    name: Windows 🪟 (PVS Studio)
```
```suggestion
    name: Windows 🪟 (PVS-Studio)
```

RytoEX reviewed (changes requested):
"PVS-Studio" is apparently the proper name of the product. SARIF is an acronym, so let's treat it that was in descriptive text to avoid confusion between that and "serif".

RytoEX commented on .github/actions/windows-analysis/action.yaml:
```diff
@@ -0,0 +1,90 @@
+name: Run PVS-Studio analysis
```
```suggestion
name: Run PVS-Studio Analysis
```

RytoEX commented on .github/actions/windows-analysis/action.yaml:
```diff
@@ -0,0 +1,90 @@
+runs:
+  using: composite
+  steps:
+    - name: 'Setup PVS-Studio'
```
```suggestion
    - name: Setup PVS-Studio
```

RytoEX commented on .github/actions/windows-analysis/action.yaml:
```diff
@@ -0,0 +1,90 @@
+        . ${env:GITHUB_ACTION_PATH}\Invoke-External.ps1
+        Invoke-External "C:\Program Files (x86)\PVS-Studio\PVS-Studio_Cmd.exe" credentials -u ${{ inputs.pvsUsername }} -n ${{ inputs.pvsKey }}
+
+    - name: Run PVS Analysis
```
```suggestion
    - name: Run PVS-Studio Analysis
```

Not sure if there is a distinction here.

RytoEX commented on .github/actions/windows-analysis/action.yaml:
```diff
@@ -0,0 +1,90 @@
+        )
+        Invoke-External "C:\Program Files (x86)\PVS-Studio\PlogConverter.exe" @conversionParams
+
+    - name: 'Upload PVS Logs'
```
```suggestion
    - name: Upload PVS Logs
```

RytoEX commented on .github/actions/windows-analysis/action.yaml:
```diff
@@ -0,0 +1,90 @@
+          ${{ github.workspace }}/analysis.plog
+          ${{ github.workspace }}/analysis.plog.sarif
+
+    - name: 'Upload PVS Report'
```
```suggestion
    - name: Upload PVS Report
```

RytoEX reviewed (changes requested):
Nits, but otherwise seems fine.
